src/hotspot/share/runtime/mountUnmountDisabler.cpp

```diff
@@ -0,0 +1,430 @@
+    _is_self(false)
+{
+  if (!Continuations::enabled()) {
+    return; // MountUnmountDisabler is no-op without virtual threads
+  }
+  if (Thread::current_or_null() == nullptr) {
+    return;  // Detached thread, can be a call from Agent_OnLoad.
+  }
+  assert(!JavaThread::current()->is_in_VTMS_transition(), "");
+  VTMS_transition_disable_for_all();
+}
+
+MountUnmountDisabler::~MountUnmountDisabler() {
+  if (!Continuations::enabled()) {
+    return; // MountUnmountDisabler is a no-op without virtual threads
+  }
+  if (Thread::current_or_null() == nullptr) {
+    return;  // Detached thread, can be a call from Agent_OnLoad.
+  }
+  if (_is_self) {
+    return; // no need for current thread to disable and enable transitions for itself
+  }
+  if (_vthread() != nullptr) {
+    VTMS_transition_enable_for_one(); // enable VTMS transitions for one virtual thread
+  } else {
+    VTMS_transition_enable_for_all(); // enable VTMS transitions for all virtual threads
+  }
+}
+
+// disable VTMS transitions for one virtual thread
+void
+MountUnmountDisabler::VTMS_transition_disable_for_one() {
+  MonitorLocker ml(VTMSTransition_lock);
+  while (exclusive_operation_ongoing()) {
+    ml.wait(10);
+  }
+
+  inc_active_disablers();
+  java_lang_Thread::inc_VTMS_transition_disable_count(_vthread());
+
+  // Prevent load of transition flag from floating up.
+  OrderAccess::storeload();
+
+  while (java_lang_Thread::is_in_VTMS_transition(_vthread())) {
+    ml.wait(10); // wait while the virtual thread is in transition
+  }
+
+  // Start of the critical region. Prevent future memory
+  // operations to be ordered before we read the transition flag.
+  // This matches the release fence in end_transition().
+  OrderAccess::acquire();
+  DEBUG_ONLY(JavaThread::current()->set_is_VTMS_transition_disabler(true);)
+}
+
+// disable VTMS transitions for all virtual threads
+void
+MountUnmountDisabler::VTMS_transition_disable_for_all() {
+  JavaThread* thread = JavaThread::current();
+
+  MonitorLocker ml(VTMSTransition_lock);
+  while (exclusive_operation_ongoing()) {
+    ml.wait(10);
+  }
+  if (_is_exclusive) {
+    set_exclusive_operation_ongoing(true);
+    while (active_disablers() > 0) {
+      ml.wait(10);
+    }
+  }
+  inc_active_disablers();
+  inc_global_start_transition_disable_count();
+
+  // Prevent loads of transition flag from floating up. Technically not
+  // required since JavaThreadIteratorWithHandle includes full fence.
+  OrderAccess::storeload();
+
+  // Block while some mount/unmount transitions are in progress.
+  // Debug version fails and prints diagnostic information.
+  for (JavaThreadIteratorWithHandle jtiwh; JavaThread *jt = jtiwh.next(); ) {
+    while (jt->is_in_VTMS_transition()) {
+      ml.wait(10);
+    }
+  }
+
+  // Start of the critical region. Prevent future memory
+  // operations to be ordered before we read the transition flags.
+  // This matches the release fence in end_transition().
+  OrderAccess::acquire();
```

Comment:
Surely the use of the iterator already provides the necessary ordering guarantee here as well. ?